src/Microsoft.Diagnostics.WebSocketServer/WebSocketServer.cs

```diff
@@ -0,0 +1,296 @@
+            _objLock = new SemaphoreSlim(1, 1);
+        }
+
+        public async Task Enqueue(T t, CancellationToken cancellationToken)
+        {
+            bool locked = false;
+            try
+            {
+                await _empty.WaitAsync(cancellationToken);
+                await _objLock.WaitAsync(cancellationToken);
+                locked = true;
+                _obj = t;
+            }
+            finally
+            {
+                if (locked)
+                {
+                    _objLock.Release();
+                    _full.Release();
+                }
+            }
+        }
+
+        public async Task<T> Dequeue(CancellationToken cancellationToken)
+        {
+            bool locked = false;
+            try
+            {
+                await _full.WaitAsync(cancellationToken);
+                await _objLock.WaitAsync(cancellationToken);
+                locked = true;
+                T t = _obj;
+                _obj = default;
+                return t;
+            }
+            finally
+            {
+                if (locked)
+                {
+                    _objLock.Release();
+                    _empty.Release();
+                }
+            }
+        }
+
+    }
+
+    internal class Conn
+    {
+        private readonly WebSocket _webSocket;
+        private readonly HttpContext _context;
+        private readonly TaskCompletionSource _streamDisposed;
+        public Conn(HttpContext context, WebSocket webSocket, TaskCompletionSource streamDisposed)
+        {
+            _context = context;
+            _webSocket = webSocket;
+            _streamDisposed = streamDisposed;
+        }
+        public Stream GetStream()
+        {
+            return new WebSocketStreamAdapter(_webSocket, OnStreamDispose);
+        }
+
+        private void OnStreamDispose()
+        {
+            _streamDisposed.SetResult();
+        }
+    }
+}
+
+public interface IWebSocketConnectionHandler
+{
+    Task Handle(HttpContext context, WebSocket webSocket, CancellationToken cancellationToken);
+}
+
+public class WebSocketServer
+{
+    public record Options
+    {
+        public string Scheme { get; set; } = "http";
+        public string Host { get; set; } = default;
+        public string Path { get; set; } = default!;
+        public string Port { get; set; } = default;
+
+
+        public void Assign(Options other)
+        {
+            Scheme = other.Scheme;
+            Host = other.Host;
+            Port = other.Port;
+            Path = other.Path;
+        }
+    }
+
+    private readonly IHost _host;
+    private WebSocketServer(IHost host)
+    {
+        _host = host;
+    }
+
+    private static string[] MakeUrls(string scheme, string host, string port) => new string[] { $"{scheme}://{host}:{port}" };
+    public static WebSocketServer CreateWebServer(Options options, Func<HttpContext, WebSocket, CancellationToken, Task> connectionHandler)
```

Comment:
@radical this is inspired by the console redirection WebSocket code (an early version of it, I haven't looked at what's been done to it lately).  Could you take a look?

Also, if I provided some middleware that the wasm AppHost could consume in order to do the diagnostic websocket support in-proc, what should the interface look like?

Reply:
Ok, I'll hold off on refactoring the diagnostic socket stuff until a future PR.